streaming/stream_transfer_task.cc

```diff
@@ -203,15 +203,27 @@ future<> send_mutation_fragments(lw_shared_ptr<send_info> si) {
         }();
 
         auto sink_op = [sink, si, got_error_from_peer] () mutable -> future<> {
-            return do_with(std::move(sink), [si, got_error_from_peer] (rpc::sink<frozen_mutation_fragment, stream_mutation_fragments_cmd>& sink) {
-                return repeat([&sink, si, got_error_from_peer] () mutable {
-                    return si->reader(db::no_timeout).then([&sink, si, s = si->reader.schema(), got_error_from_peer] (mutation_fragment_opt mf) mutable {
-                        if (mf && !(*got_error_from_peer)) {
+            bool partition_opened = false;
+            return do_with(std::move(sink), partition_opened, [si, got_error_from_peer] (rpc::sink<frozen_mutation_fragment, stream_mutation_fragments_cmd>& sink, bool& partition_opened) {
+                return repeat([&sink, &partition_opened, si, got_error_from_peer] () mutable {
+                    return si->reader(db::no_timeout).then([&sink, &partition_opened, si, s = si->reader.schema(), got_error_from_peer] (mutation_fragment_opt mf) mutable {
+                        if (*got_error_from_peer) {
+                            return make_exception_future<stop_iteration>(std::runtime_error("Got status error code from peer"));
+                        }
+                        if (mf) {
+                            if (mf->is_partition_start()) {
+                                partition_opened = true;
+                            } else if (mf->is_end_of_partition()) {
+                                partition_opened = false;
+                            }
                             frozen_mutation_fragment fmf = freeze(*s, *mf);
                             auto size = fmf.representation().size();
                             streaming::get_local_stream_manager().update_progress(si->plan_id, si->id.addr, streaming::progress_info::direction::OUT, size);
                             return sink(fmf, stream_mutation_fragments_cmd::mutation_fragment_data).then([] { return stop_iteration::no; });
                         } else {
+                            if (partition_opened) {
```

Comment:
... and replace this with:
```c++
if (last_mutation_fragment_kind != mutation_fragment::kind::partition_end) {
```

Reply:
Such check would be wrong if the reader emits zero partition, no?

Reply:
Not if you initialize `last_mutation_fragment_kind` to `mutation_fragment::kind::partition_end`.

Reply:
Yes, this should work. Sounds a bit weird to me though because we do not get partition_end from the reader.

Reply:
It is weird on first read, but if you think about it, if you haven't read anything yet, you are positioned before a partition start, and `partition_end` is the only position which expresses this.